src/metabase/api/dashboard.clj

```diff
@@ -258,6 +289,35 @@
         add-query-average-durations
         (api/present-in-trash-if-archived-directly (collection/trash-collection-id)))))
 
+(def ^:private get-dashboard-fn
+  (memoize/ttl (fn [dashboard-load-id]
+                 (if dashboard-load-id
+                   (memoize get-dashboard*) ; If dashboard-load-id is set, return a memoized get-dashboard*.
+                   get-dashboard*))         ; If unset, just call through to get-dashboard*.
+               :ttl/threshold dashboard-load-cache-ttl))
+
+(def ^:private dashboard-load-metadata-provider-cache
+  (memoize/ttl (fn [_dashboard-load-id]
+                 (atom (cache/basic-cache-factory {})))
+               :ttl/threshold dashboard-load-cache-ttl))
+
+(defn- do-with-dashboard-load-id [dashboard-load-id body-fn]
+  (if dashboard-load-id
+    (binding [*dashboard-load-id*                        dashboard-load-id
+              lib.metadata.jvm/*metadata-provider-cache* (dashboard-load-metadata-provider-cache dashboard-load-id)]
+      (body-fn))
+    (body-fn)))
+
+(defmacro ^:private with-dashboard-load-id [dashboard-load-id & body]
+  `(do-with-dashboard-load-id ~dashboard-load-id (^:once fn* [] ~@body)))
+
+(defn- get-dashboard
```

Comment:
Nothing wrong with it here, but I wonder if this function should sit somewhere else for re-use as well. I think model.dashboard makes some sense, or a new namespace since both model.dashboard and api.dashboard are quite heavy namespaces.